metronome/networking/src/io/iohk/metronome/networking/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,339 @@
+package io.iohk.metronome.networking
+
+import cats.effect.concurrent.Deferred
+import cats.effect.implicits._
+import cats.effect.{Concurrent, ContextShift, Resource, Sync, Timer}
+import cats.implicits._
+import io.iohk.metronome.networking.ConnectionHandler.HandledConnection
+import io.iohk.metronome.networking.EncryptedConnectionProvider.ConnectionError
+import io.iohk.metronome.networking.RemoteConnectionManager.{
+  ConnectionAlreadyClosedException,
+  MessageReceived
+}
+import monix.catnap.ConcurrentQueue
+import monix.eval.{TaskLift, TaskLike}
+import monix.reactive.Observable
+import monix.tail.Iterant
+import scodec.Codec
+
+import java.net.InetSocketAddress
+import java.util.concurrent.{ThreadLocalRandom, TimeUnit}
+import scala.concurrent.duration.FiniteDuration
+
+class RemoteConnectionManager[F[_]: Sync, K, M: Codec](
+    connectionHandler: ConnectionHandler[F, K, M],
+    localInfo: (K, InetSocketAddress)
+) {
+
+  def getLocalInfo: (K, InetSocketAddress) = localInfo
+
+  def getAcquiredConnections: F[Set[K]] = {
+    connectionHandler.getAllActiveConnections
+  }
+
+  def incomingMessages: Iterant[F, MessageReceived[K, M]] =
+    connectionHandler.incomingMessages
+
+  def sendMessage(recipient: K, message: M): F[Unit] = {
+    connectionHandler.getConnection(recipient).flatMap {
+      case Some(connection) =>
+        //Connections could be closed by remote without us noticing, close it on our side and return error to caller
+        connection.sendMessage(message).handleErrorWith { e =>
+          //Todo logging
+          connection
+            .close()
+            .flatMap(_ =>
+              Sync[F].raiseError(ConnectionAlreadyClosedException(recipient))
+            )
+        }
+      case None =>
+        Sync[F].raiseError(ConnectionAlreadyClosedException(recipient))
+    }
+  }
+}
+//TODO add logging
+object RemoteConnectionManager {
+  case class ConnectionAlreadyClosedException[K](key: K)
+      extends RuntimeException(
+        s"Connection with node ${key}, has already closed"
+      )
```

Comment:
```suggestion
      extends RuntimeException(
        s"Connection with node ${key}, has already closed"
      ) with NoStackTrace
```
I don't think we need a huge stack trace when we log this one, it's obvious where it's coming from.